learning_observer/learning_observer/lms_integration.py

```diff
@@ -0,0 +1,287 @@
+    if lms_name == 'google' and constants.AUTH_HEADERS not in request:
+        raise aiohttp.web.HTTPUnauthorized(text="Please log in")
+    
+    # Construct the full URL using the base URL if provided, otherwise use the target URL directly
+    if base_url:
+        url = base_url + target_url.format(**kwargs)
+    else:
+        url = target_url.format(**kwargs)
+
+    # Generate a unique cache key based on the service, user, and request URL
+    cache_key = f"raw_{lms_name}/" + learning_observer.auth.encode_id('session', user[constants.USER_ID]) + '/' + learning_observer.util.url_pathname(url)
+    
+    # Check cache and return cached response if available
+    if settings.feature_flag(f"use_{lms_name}_ajax") is not None:
+        value = await cache[cache_key]
+        if value is not None:
+            # Translate keys if the service is Google, otherwise return raw JSON
+            if lms_name == 'google':
+                return learning_observer.util.translate_json_keys(
+                    json.loads(value),
+                    learning_observer.google.GOOGLE_TO_SNAKE
+                )
+            else: 
+                return json.loads(value)
+
+    # Make the actual AJAX call to the service
+    async with aiohttp.ClientSession(loop=request.app.loop) as client:
+        try:
+            async with client.get(url, headers=headers[lms_name]) as resp:
+                response = await resp.json()
+
+                # Log the AJAX request and response
+                learning_observer.log_event.log_ajax(target_url, response, request)
+                # Cache the response if the feature flag is enabled
+                if settings.feature_flag(f"use_{lms_name}_ajax") is not None:
+                    await cache.set(cache_key, json.dumps(response, indent=2))
+                # Translate keys if the service is Google, otherwise return raw JSON
+                if lms_name == 'google':
+                    return learning_observer.util.translate_json_keys(
+                        response,
+                        learning_observer.google.GOOGLE_TO_SNAKE
+                    )
+                else:
+                    resp.raise_for_status()
+                    return response
+        # Handle 401 errors for Canvas with an optional retry
+        except aiohttp.ClientResponseError as e:
+            if lms_name == 'canvas' and e.status == 401 and retry:
+                new_tokens = await learning_observer.auth.social_sso._canvas(request)
+                if 'access_token' in new_tokens:
+                    return await raw_ajax(runtime, target_url, lms_name, base_url, **kwargs)
+            raise
+
+async def raw_google_ajax(runtime, target_url, **kwargs):
+    return await raw_ajax(runtime, target_url, 'google', **kwargs)
+
+async def raw_canvas_ajax(runtime, target_url, **kwargs):
+    default_server = settings.pmss_settings.default_server(types=['lms', 'canvas_oauth'])
+    base_url = f'https://{default_server}/api/v1/'
```

Comment:
Perhaps naming this `lms_api` or something better instead of `default_server` would be more helpful. This also ties in with my prior comment about default server being ambiguous.

Reply:
Renamed to `lms_api`